src/firmware_new/boards/frankie_v1/udp_multicast.c

```diff
@@ -0,0 +1,123 @@
+static void blocking_udp_multicast_loop(void *arg);
+
+struct multicast_config
+{
+    ip_addr_t multicast_address;
+    unsigned multicast_port;
+    unsigned send_port;
+} typedef multicast_config_t;
+
+/*
+ * Thread that creates a send and recv socket, joins the specified
+ * multicast group and start listening for packets. Increments a counter
+ * on every multicast packet and sends a unicast packet back to the sender.
+ *
+ */
+static void blocking_udp_multicast_loop(void *arg)
+{
+    multicast_config_t *config = (multicast_config_t *)arg;
+    struct netbuf *rx_buf      = NULL;
+    struct netbuf *tx_buf      = NULL;
+    int msg_count              = 0;
+
+    // TODO proper err handling, for now all errs are ignored
+    // https://github.com/UBC-Thunderbots/Software/issues/1190
+    err_t err;
+
+    // create two new UDP connections on the heap
+    // NOTE: we need two seperate UDP sockets as we will be
+    // receiving multicast but sending unicast packets to avoid
+    // network congestion (peer robots don't need to know the status
+    // of other robots)
+    struct netconn *recvconn = netconn_new(NETCONN_UDP_IPV6);
+    struct netconn *sendconn = netconn_new(NETCONN_UDP_IPV6);
+
+    // Bind the socket to the multicast address and port
+    // we then use that connection profile to join the specified
+    // multicast group.
+    //
+    // For the send socket, we use IP6_ADDR_ANY
+    // which will default to the ethernet interface on the STM32H7
+    //
+    // This will remain the same regardless of communicating over ethernet
+    // or WiFi because both of those media use the ETH interface
+    netconn_bind(recvconn, &config->multicast_address, config->multicast_port);
+    netconn_bind(sendconn, IP6_ADDR_ANY, config->send_port);
+    netconn_join_leave_group(recvconn, &config->multicast_address, NULL, NETCONN_JOIN);
+
+    while (1)
+    {
+        err = netconn_recv(recvconn, &rx_buf);
+
+        if (err == ERR_OK)
+        {
+            tx_buf = netbuf_new();
+            netbuf_alloc(tx_buf, robot_ack_size);
+
+            // Create a stream that reads from the buffer
+            pb_istream_t in_stream =
+                pb_istream_from_buffer((uint8_t *)rx_buf->p->payload, rx_buf->p->tot_len);
+
+            if (pb_decode(&in_stream, control_msg_fields, &control))
+            {
+                // update proto
+                ack.msg_count = msg_count++;
+
+                // serialize proto
+                pb_ostream_t stream = pb_ostream_from_buffer(buffer, sizeof(buffer));
+                pb_encode(&stream, robot_ack_fields, &ack);
+
+                // package payload and send over udp
+                tx_buf->p->payload = buffer;
+                netconn_sendto(sendconn, tx_buf, (const ip_addr_t *)&(rx_buf->addr),
+                               config->send_port);
+            }
+            netbuf_delete(tx_buf);
+        }
+        netbuf_delete(rx_buf);
+    }
+}
```

Comment:
Nice, this function is pretty clean! This will probably end up being it's own RTOS task running at some fixed frequency (which should alleviate the hard faults we saw at the end of Saturday)

Reply:
We would also need to add some sort of "API" to allow for any thread to call send_proto and send a message so we don't respond to every msg we get.